[KOGITO-2285] Shared libaries: Git tagging

Kevin-Mok opened the pull request:
See: https://issues.redhat.com/browse/KOGITO-2285

This PR adds a `tagRepository` function as per the issue. The core of the code is basically the same as [this CloudBees article](https://support.cloudbees.com/hc/en-us/articles/360027646491-Pipeline-Equivalent-to-Git-Publisher).  I originally used SSH for authentication but realized the existing pipelines used HTTPS, so this also uses HTTPS authentication. The structure of the function mimics `mergeSourceIntoTarget` with using the `checkout` and `resolveRepository` functions to checkout the repository and also annotates progress with info/error messages.

### Example Usage
```Jenkinsfile
stage('Tag kogito-runtimes') {
    steps {
        script {
            githubscm.tagRepository('kogito-runtimes', 'kiegroup', 'master', 'Example User', 'example@example.com', 'v2.00')
        }
    }
}
```

Ginxo commented on vars/githubscm.groovy:
```diff
@@ -74,6 +74,38 @@ def mergeSourceIntoTarget(String repository, String sourceAuthor, String sourceB
+    """)
+
+    try {
+        withCredentials([usernamePassword(credentialsId: 'kie-ci, usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]){
```
there's a typo here `credentialsId: 'kie-ci` should be `credentialsId: 'kie-ci'`

Kaitou786 commented:
@Ginxo I think both pushCommit will use different `withCredentials` as in Kevin's case he will be pushing the commit in the kie organization repositories, but in my case I want to push the commit to another org, In my case the push is dependent on setting the `GITHB_TOKEN` env (Which is also used for functions like fork and PR) and does not uses the `withCredentials` plugin.

radtriste commented:
@Ginxo @Kevin-Mok I like the idea of single responsibility of methods. Then it is up to the "user" to decide the course of actions :) I would also make the `origin` repository as a parameter

@Kaitou786 credentials could be an optional parameter of the `pushCommit` method. btw note that we may reuse a common bsig github group for those actions, like pushing. And so, the current `kie-ci` may be enough

Kaitou786 commented:
@radtriste I think than we can have the username and password as the parameter to the pushCommit method, and use the `withCrdentials` in the pipeline step to retrieve these username and password and pass them as parameter to the function

Ginxo commented:
@Kaitou786 I agree with @radtriste I would prefer his approach doing
```
def pushCommit(String target, String credentialsId = 'kie-ci') {
    def currentCommit = getCommit()
    try {
        withCredentials([usernamePassword(credentialsId: credentialsId, usernameVariable: 'GIT_USERNAME', passwordVariable: 'GIT_PASSWORD')]){
            sh("""
                git config --local credential.helper "!f() { echo username=\\$GIT_USERNAME; echo password=\\$GIT_PASSWORD; }; f"
                git push origin ${target}
            """)
        }
    } catch (Exception e) {
        println "[ERROR] Can't push existing '${target}' to remote."
        throw e;
    }
    println "[INFO] Pushed commit '${currentCommit}' to remote '${target}'."
}
```

Kaitou786 commented:
I was think maybe some thing like
 ```void pushCommit(String branchName, String githubUsername, String githubPassword){
    sh "git config --global push.default simple"
    sh """
    git config --local credential.helper "!f() { echo username="${githubUsername}"; echo password="${githubPassword}"; }; f"
    """
    sh "git push origin ${branchName} --force"
}```

Ginxo commented:
@Kaitou786 I still thinking to wrap it with `withCredentials` is a better approach since is the proper way to do it in Jenkins (I mean to create credentials and then use those credentials) but I'm not an expert, just an opinion

Kevin-Mok commented:
## Reasoning for Single Function

> // Not sure if this is out of scope for this method ¿?
> // checkout(resolveRepository(repository, author, branch, false)) 

@Ginxo I originally had the method so that it wouldn't clone the repository such that you would use the `tagRepository` function as such:

```
checkoutIfExists(repository, author, branches, defaultAuthor, defaultBranches, mergeTarget)
tagRepository(tagUserName, tagUserEmail, tagName)
```

But, I thought it'd be simpler to use if it was combined into just one function.  I do agree that it'd be good for the push function to be extracted into another method so that that functionality can be reused for other purposes, like @Kaitou786's task.

## Changes
### Extracted pushing to `pushObject()`
I removed checking out the repository from the function and extracted the pushing to `pushObject()` (object [since any SHA or tag can be pushed](https://git-scm.com/docs/git-push#Documentation/git-push.txt-ltrefspecgt82308203)).  It allows for any remote to be pushed to as @radtriste suggested as well as taking in any `credentialsId` as @Ginxo suggested.

### Optional `buildTag` parameter
I added an optional `buildTag` parameter to `tagRepository()` to improve the tag message. Originally the tag message was just "`Tagging ${tagName}`" to copy [the message from the runtimes script](https://github.com/kiegroup/kogito-runtimes/blob/13216df6ea79dd77e3baaf42aaba0bd80ab85367/scripts/release/05-tag-and-push-local-release-branches.sh#L52 ). But, I thought it would be more useful/descriptive if it had the build tag, so that the message would be "`Tagged by Jenkins in build 'jenkins-Custom-kmok-github-release-test-16'`".  This is an optional parameter, so it doesn't have to be used; in that case, it would just be "`Tagged by Jenkins.`".

### New Usage
An example tagging stage would be as follows:

```Jenkinsfile
stage('Tag repo') {
    steps {
        script {
            githubscm.checkoutIfExists('github-release-test',
                "Kevin-Mok", "master", 'kiegroup', "master", false)
            githubscm.tagRepository('Test', 'test@example.com', "$BUILD_TAG", "$BUILD_TAG")
            githubscm.pushObject('origin', "$BUILD_TAG", "jenkins-kogito")
        }
    }
}
```

[Link to example console log](https://jenkins-kogito-tools.apps.kogito.automation.rhmw.io/job/Custom/job/kmok/job/github-release-test/19/console).

Ginxo commented:
@radtriste please review again after @Kevin-Mok changes, thanks guys

Ginxo commented on vars/githubscm.groovy:
```diff
@@ -125,6 +124,44 @@ def mergePR(String pullRequestLink, String credentialID='kie-ci') {
+    sh("""
+        git config user.name '${tagUserName}'
+        git config user.email '${tagUserEmail}'
+        git tag -a '${tagName}' -m '${tagMessage}'
```
In case you send a `buildTag` your message will be something like 
`'Tagged by Jenkins in build 'buildTag'.'` this double quote will be an error (I guess)
Probably
`def tagMessageEnding = buildTag ? " in build \"${buildTag}\"." : '.'` will be more appropriated

Kevin-Mok replied:
Good catch. Fixed, thanks.
